### PR TITLE
CLOUDP-260768: Plugin install command

### PR DESCRIPTION
**Pull request by stefan4h**
## Proposed changes

- create `plugin install` command
- create `plugin_asset` file that contains functions that can be used for `install` and `update`

_Jira ticket:_ CLOUDP-260768

### CODE REVIEW

File: docs/command/atlas-plugin-install.txt [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+   :class: singlecol
+
+Install Atlas CLI plugin from a GitHub repository.
+
+Install an Atlas CLI plugin from a GitHub repository.
+The GitHub repository can be specified using either the "<github-owner>/<github-repository-name>" format or a full URL.
+By default, the latest release on GitHub will be used to install the plugin.
+If a specific version is needed, it can be specified using the --version flag.
+
+An example plugin can be found here: https://github.com/mongodb/atlas-cli-plugin-example
```
Comment (JuliaMongo) [inline docs/command/atlas-plugin-install.txt:22]: MongoDB provides an [example plugin](https://github.com/mongodb/atlas-cli-plugin-example).

File: docs/command/atlas-plugin-install.txt [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+     - false
+     - Name of the profile to use from your configuration file. To learn about profiles for the Atlas CLI, see https://dochub.mongodb.org/core/atlas-cli-save-connection-settings.
+
+Examples
+--------
+
+.. code-block::
+   :copyable: false
+
+   # Install latest version of plugin:
```
Comment (JuliaMongo) [inline docs/command/atlas-plugin-install.txt:89]: Install the latest version of the plugin:

File: docs/command/atlas-plugin-install.txt [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+   atlas plugin install mongodb/atlas-cli-plugin-example
+   atlas plugin install https://github.com/mongodb/atlas-cli-plugin-example
+   
+   
+.. code-block::
+   :copyable: false
+
+   # Install a specific version of plugin:
```
Comment (JuliaMongo) [inline docs/command/atlas-plugin-install.txt:97]: the plugin

**Review by JuliaMongo — APPROVED** [review]
LGTM with a few editing suggestions. Basically asking to please use active voice and to avoid using links with "here". Thank you!

File: internal/cli/plugin/plugin_asset.go [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+	name := parts[len(parts)-1]
+
+	if owner == "" || name == "" {
+		return "", "", err
+	}
+
+	return owner, name, nil
+}
+
+func parseReleaseVersion(arg string) (string, error) {
```
Comment (jeroenvervaeke) [inline internal/cli/plugin/plugin_asset.go:130]: Why not return a version here instead of a string?
You return an empty string for latest anyway.

File: internal/cli/plugin/plugin_asset.go [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+	pluginZipFilePath := filepath.Join(pluginsDefaultDirectory, "zippedPluginFiles.tar.gz")
+	pluginTarGzFile, err := os.Create(pluginZipFilePath)
+	if err != nil {
+		return "", errors.New("could not open plugins directory")
+	}
+	defer pluginTarGzFile.Close()
+
+	_, err = io.Copy(pluginTarGzFile, rc)
+	if err != nil {
+		return "", errors.New("failed to save asset to plugin directory: " + err.Error())
```
Comment (jeroenvervaeke) [inline internal/cli/plugin/plugin_asset.go:230]: Don't forget to clean up the file you made at line 222

File: internal/cli/plugin/plugin_asset.go [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+	}
+	defer file.Close()
+
+	gzipReader, err := gzip.NewReader(file)
+	if err != nil {
+		return fmt.Errorf("failed to create gzip reader: %w", err)
+	}
+	defer gzipReader.Close()
+
+	tarReader := tar.NewReader(gzipReader)
```
Comment (jeroenvervaeke) [inline internal/cli/plugin/plugin_asset.go:166]: [q] do we need to close the tar reader?

File: internal/cli/plugin/plugin_asset.go [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+		if strings.HasPrefix(fileName, "..") {
+			return fmt.Errorf("illegal file path for extracted plugin asset file: %s", fileName)
+		}
+
+		filePath := filepath.Join(dest, fileName)
+
+		switch header.Typeflag {
+		case tar.TypeDir:
+			if err := os.MkdirAll(filePath, os.FileMode(header.Mode)); err != nil {
+				return errors.New("failed to create directory while extracting plugin asset")
```
Comment (jeroenvervaeke) [inline internal/cli/plugin/plugin_asset.go:187]: Can you turn all these errors in consts and define them on the top of this file?

File: internal/plugin/plugin.go [hunk 8]
```diff
@@ -15,28 +15,35 @@
...
 	// Load manifests from plugin directories
-	if loadedManifests, err := getManifestsFromPluginDirectory("./plugins"); err != nil {
-		logPluginWarning(`could not load manifests from directory "./plugins" because of error: %s`, err.Error())
-	} else {
-		manifests = append(manifests, loadedManifests...)
+	if defaultPluginDirectory, err := GetDefaultPluginDirectory(); err == nil {
+		if loadedManifests, err := getManifestsFromPluginsDirectory(defaultPluginDirectory); err != nil {
+			logPluginWarning(`could not load manifests from directory "%s" because of error: %s`, defaultPluginDirectory, err.Error())
+		} else {
+			manifests = append(manifests, loadedManifests...)
```
Comment (jeroenvervaeke) [inline internal/plugin/plugin.go:41]: What's the point of this else branch? Can there still be loadedManifests when there's an error?

File: internal/plugin/plugin_manifest.go [hunk 9]
```diff
@@ -92,39 +107,49 @@ func getManifestsFromPluginDirectory(pluginDirectory string) ([]*Manifest, error
 			continue
 		}
 
-		pluginDirectoryPath := fmt.Sprintf("%s/%s", pluginDirectory, directory.Name())
-		manifestFileData, err := getManifestFileBytes(pluginDirectoryPath)
+		pluginDirectoryPath := fmt.Sprintf("%s/%s", pluginsDirectory, directory.Name())
```
Comment (jeroenvervaeke) [inline internal/plugin/plugin_manifest.go:110]: path.join

File: tools/docs/main.go [hunk 10]
```diff
@@ -42,6 +43,7 @@ func main() {
 
 	atlasBuilder := root.Builder()
 	atlasBuilder.InitDefaultCompletionCmd()
+	atlasBuilder.AddCommand(plugin.Builder(nil, nil))
```
Comment (jeroenvervaeke) [inline tools/docs/main.go:46]: Why did you remove this function from the root builder?

**Review by jeroenvervaeke — CHANGES_REQUESTED** [review]

File: internal/cli/plugin/install.go [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+
+			version, err := parseReleaseVersion(args[0])
+			if err != nil {
+				return err
+			}
+			opts.releaseVersion = version
+
+			opts.ghClient = github.NewClient(nil)
+
+			assets, err := opts.getPluginAssets()
```
[the remark this answers is not in the document]
Comment (stefan4h) [reply internal/cli/plugin/install.go:149]: I used the `PreRunE` for setting up the `InstallOpts` with all the necessary values that will then be used in the `RunE` to actually do the download and install of the plugin. 
I can see how the name `getPluginAssets` is a bit misleading as it does not actually get any assets, just the asset information. I changed the name now to `getPluginAssetInfo`.

File: internal/cli/plugin/plugin_asset.go [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+
+	parts := strings.Split(arg, "/")
+
+	err := errors.New(`github parameter is invalid. It needs to have the format "<github-owner>/<github-repository-name>"`)
+
+	minParts := 2
+	if len(parts) < minParts {
+		return "", "", err
+	}
+	owner := parts[len(parts)-2]
```
[the remark this answers is not in the document]
Comment (stefan4h) [reply internal/cli/plugin/plugin_asset.go:120]: If we want to support the option for the user to install a plugin like this `atlas plugin install https://github.com/mongodb/atlas-cli-plugin-example` or like this `atlas plugin install github.com/mongodb/atlas-cli-plugin-example` then the parts slice can be longer than 2 but we only care about the last two values.
Or am I missing something here?

File: internal/cli/plugin/plugin_asset.go [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+func (opts *AssetOpts) fullRepositoryDefinition() string {
+	return fmt.Sprintf("%s/%s", opts.repositoryOwner, opts.repositoryName)
+}
+
+func (opts *AssetOpts) getPluginAssets() ([]*github.ReleaseAsset, error) {
+	var err error
+	var release *github.RepositoryRelease
+
+	// download latest release if version is not specified
+	if opts.releaseVersion == "" {
```
[the remark this answers is not in the document]
Comment (stefan4h) [reply internal/cli/plugin/plugin_asset.go:54]: change the type from string to Version now. If the Version is "nil", the latest release will be used

File: internal/cli/plugin/plugin_asset.go (hunk 6 above)
Comment (stefan4h) [reply internal/cli/plugin/plugin_asset.go:166]: No, ther tarReader does not have a `Close` function

File: tools/docs/main.go (hunk 10 above)
Comment (stefan4h) [reply tools/docs/main.go:46]: Calling the RegisterPlugin function in the `root.Builder()` will cause the unit tests for the builder to fail if there are any plugins installed because that will change the amount of sub commands of the rootCmd.

File: internal/cli/plugin/install.go [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+)
+
+type InstallOpts struct {
+	cli.GlobalOpts
+	cli.OutputOpts
+	AssetOpts
+	plugins []*plugin.Plugin
+}
+
+func (opts *InstallOpts) validateForExistingPlugins() error {
```
[the remark this answers is not in the document]
Comment (stefan4h) [reply internal/cli/plugin/install.go]: changed it to `checkForDuplicatePlugins `

File: internal/cli/plugin/plugin_asset.go [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+	rc, _, err := opts.ghClient.Repositories.DownloadReleaseAsset(context.Background(), opts.repositoryOwner, opts.repositoryName, assetID, http.DefaultClient)
+
+	if err != nil {
+		return nil, fmt.Errorf("could not download asset with ID %d from %s", assetID, opts.fullRepositoryDefinition())
+	}
+
+	return rc, nil
+}
+
+func parseGithubValues(arg string) (string, string, error) {
```
[the remark this answers is not in the document]
Comment (stefan4h) [reply internal/cli/plugin/plugin_asset.go]: done

File: internal/cli/plugin/install.go [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+}
+
+func (opts *InstallOpts) Run() error {
+	rc, err := opts.getPluginAssetAsReadCloser()
+	if err != nil {
+		return err
+	}
+
+	pluginZipFilePath, err := saveReadCloserToPluginAssetZipFile(rc)
+	defer os.Remove(pluginZipFilePath)
```
[the remark this answers is not in the document]
Comment (stefan4h) [reply internal/cli/plugin/install.go]: yes, changed that now

File: internal/cli/plugin/install.go [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+	if err != nil {
+		return err
+	}
+
+	pluginDirectoryPath, err := opts.extractPluginAssetZipFile(pluginZipFilePath)
+	if err != nil {
+		if pluginDirectoryPath != "" {
```
[the remark this answers is not in the document]
Comment (stefan4h) [reply internal/cli/plugin/install.go]: fixed that now

File: internal/cli/plugin/plugin_asset.go [hunk 18]
```diff
@@ -0,0 +1,253 @@
...
+	existingCommands []*cobra.Command
+	ghClient         *github.Client
+	repositoryOwner  string
+	repositoryName   string
+	releaseVersion   string
+	pluginAssets     []*github.ReleaseAsset
+}
+
+func (opts *AssetOpts) fullRepositoryDefinition() string {
+	return fmt.Sprintf("%s/%s", opts.repositoryOwner, opts.repositoryName)
```
[the remark this answers is not in the document]
Comment (stefan4h) [reply internal/cli/plugin/plugin_asset.go]: called it `repository`

File: internal/cli/plugin/plugin_asset.go [hunk 19]
```diff
@@ -0,0 +1,253 @@
...
+	return rc, nil
+}
+
+func parseGithubValues(arg string) (string, string, error) {
+	arg = strings.Split(arg, "@")[0]
+	arg = strings.TrimSuffix(arg, "/")
+
+	parts := strings.Split(arg, "/")
+
+	err := errors.New(`github parameter is invalid. It needs to have the format "<github-owner>/<github-repository-name>"`)
```
[the remark this answers is not in the document]
Comment (stefan4h) [reply internal/cli/plugin/plugin_asset.go]: done

File: internal/cli/plugin/plugin_asset.go [hunk 20]
```diff
@@ -0,0 +1,253 @@
...
+
+	minParts := 2
+	if len(parts) < minParts {
```
[the remark this answers is not in the document]
Comment (stefan4h) [reply internal/cli/plugin/plugin_asset.go:152]: done

File: internal/cli/plugin/plugin_asset.go [hunk 21]
```diff
@@ -0,0 +1,253 @@
...
+	}
+
+	return owner, name, nil
+}
+
+func parseReleaseVersion(arg string) (string, error) {
+	parts := strings.Split(arg, "@")
+
+	minParts := 2
+	if len(parts) < minParts {
```
[the remark this answers is not in the document]
Comment (stefan4h) [reply internal/cli/plugin/plugin_asset.go]: fixed

File: internal/cli/plugin/plugin_asset.go (hunk 4 above)
Comment (stefan4h) [reply internal/cli/plugin/plugin_asset.go]: done

File: internal/cli/plugin/plugin_asset.go (hunk 5 above)
Comment (stefan4h) [reply internal/cli/plugin/plugin_asset.go]: done

File: internal/cli/plugin/plugin_asset.go (hunk 7 above)
Comment (stefan4h) [reply internal/cli/plugin/plugin_asset.go]: done

File: internal/plugin/plugin.go (hunk 8 above)
Comment (stefan4h) [reply internal/plugin/plugin.go]: changed that now

File: internal/plugin/plugin_manifest.go (hunk 9 above)
Comment (stefan4h) [reply internal/plugin/plugin_manifest.go]: fixed

File: internal/plugin/plugin.go (hunk 8 above)
Comment (jeroenvervaeke) [reply internal/plugin/plugin.go]: I'm sorry, I must have read the original code wrong.
I thought you were adding `manifests = append(manifests, loadedManifests...)` in both branches, which you were not. Can you revert this change?

Comment (stefan4h) [reply internal/plugin/plugin.go:45]: no worries, just reverted it :)

**Review by jeroenvervaeke — APPROVED** [review]
LGTM!
